### PR TITLE
MPNet: Masked and Permuted Pre-training for Language Understanding

### PR DESCRIPTION
**Pull request by StillKeepTry**
# Model addition
[MPNet](https://arxiv.org/abs/2004.09297)

## Model description
MPNet introduces a novel self-supervised objective named masked and permuted language modeling for language understanding. It inherits the advantages of both the masked language modeling  (MLM) and the permuted language modeling (PLM) to addresses the limitations of MLM/PLM, and further reduce the inconsistency between the pre-training and fine-tuning paradigms.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hey @StillKeepTry - could you maybe link the paper corresponding to your model add a small PR description? :-) That would be very helpful

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (patrickvonplaten) [inline src/transformers/models/mpnet/tokenization_mpnet.py:45]: Colud we maybe try to not inherit from `BertTokenizer` here? It seems like this class overwrites quite some methods - in this case it would be awesome if you could just copy past the required `BertTokenizer` functionality here. @LysandreJik @sgugger - what is your opinion on that?

File: src/transformers/models/mpnet/configuration_mpnet.py
Comment (patrickvonplaten) [inline src/transformers/models/mpnet/configuration_mpnet.py:25]: do we have pre-trained weights?

**Comment by patrickvonplaten** [comment]
Thanks for the new PR @StillKeepTry - could you add a `test_modeling_mpnet.py` file?

**Comment by JetRunner** [comment]
> Thanks for the new PR @StillKeepTry - could you add a `test_modeling_mpnet.py` file - it would be important to test the model :-) 
> 
> 
> 
> Also it would be amazing if you could give some context of MPNet - is there a paper, blog post, analysis, results that go along with the model? And are there pretrained weights? 
> 
> 
> 
> Thanks a lot! 

https://arxiv.org/abs/2004.09297

**Comment by StillKeepTry** [comment]
> > Thanks for the new PR @StillKeepTry - could you add a `test_modeling_mpnet.py` file - it would be important to test the model :-)
> > Also it would be amazing if you could give some context of MPNet - is there a paper, blog post, analysis, results that go along with the model? And are there pretrained weights?
> > Thanks a lot!
> 
> https://arxiv.org/abs/2004.09297

ok

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (LysandreJik) [reply src/transformers/models/mpnet/tokenization_mpnet.py:45]: Agreed, I don't see the point of inheriting from `BertTokenizer` here either.

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (LysandreJik) [inline src/transformers/models/mpnet/tokenization_mpnet.py:69]:
```suggestion
        unk_token (:obj:`string`, `optional`, defaults to "[UNK]"):
```

File: src/transformers/models/mpnet/configuration_mpnet.py
Comment (LysandreJik) [inline src/transformers/models/mpnet/configuration_mpnet.py:58]: It would be great if this could be formatted like our other configuration files. You can probably copy-paste a lot from BERT: https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/configuration_bert.py#L62-L100

File: src/transformers/models/mpnet/modeling_mpnet.py
Comment (LysandreJik) [inline src/transformers/models/mpnet/modeling_mpnet.py]: No need for this, you're importing transformers' logging utility already.

File: src/transformers/models/mpnet/modeling_mpnet.py
Comment (LysandreJik) [inline src/transformers/models/mpnet/modeling_mpnet.py:97]: Should this be placed above the method? @sgugger

File: src/transformers/models/mpnet/modeling_mpnet.py
Comment (LysandreJik) [inline src/transformers/models/mpnet/modeling_mpnet.py:469]: Could you update this to be more in line with the other models' documentation? You can copy-paste a lot from BERT's: https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/modeling_bert.py#L691-L738

File: src/transformers/models/mpnet/modeling_mpnet.py
Comment (LysandreJik) [inline src/transformers/models/mpnet/modeling_mpnet.py:935]:
```suggestion
        x = features[:, 0, :]  # take <s> token (equiv. to BERT's [CLS] token)
```

File: src/transformers/models/mpnet/modeling_tf_mpnet.py
Comment (LysandreJik) [inline src/transformers/models/mpnet/modeling_tf_mpnet.py:1]: Could you add the copyright here?

**Review by LysandreJik — COMMENTED** [review]
This is exceptionally clean, thank you! I've left a few comments regarding the documentation.

Also,  we've recently updated the way to process the inputs in TensorFlow. Maybe @jplu can give you some tips or help you implement the changes.

As @patrickvonplaten mentioned, the two test files `test_modeling_mpnet.py` and `test_modeling_tf_mpnet.py` are necessary for us to merge this. Let us know if you need help implementing those and we'll be sure to give you pointers.

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:55]:
```suggestion
        bos_token (:obj:`string`, `optional`, defaults to :obj:`"<s>"`):
```

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:58]:
```suggestion

            .. note::
            
                When building a sequence using special tokens, this is not the token that is used for the
                beginning of sequence. The token used is the :obj:`cls_token`.
```
This needs a bit more new lines to be properly formatted.

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:59]:
```suggestion
        eos_token (:obj:`string`, `optional`, defaults to :obj:`"</s>"`):
```

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:61]:
```suggestion
            The end of sequence token. 
            
            .. note::
            
                When building a sequence using special tokens, this is not the token
                that is used for the end of sequence. The token used is the :obj:`sep_token`.
```

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:62]:
```suggestion
        sep_token (:obj:`string`, `optional`, defaults to :obj:`"</s>"`):
```

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:66]:
```suggestion
        cls_token (:obj:`string`, `optional`, defaults to :obj:`"<s>"`):
```

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:69]:
```suggestion
        unk_token (:obj:`string`, `optional`, defaults to :obj:`"[UNK]"`):
```

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:72]:
```suggestion
        pad_token (:obj:`string`, `optional`, defaults to :obj:`"<pad>"`):
```

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet.py:74]:
```suggestion
        mask_token (:obj:`string`, `optional`, defaults to :obj:`"<mask>"`):
```

File: src/transformers/models/mpnet/tokenization_mpnet_fast.py
Comment (sgugger) [inline src/transformers/models/mpnet/tokenization_mpnet_fast.py:59]: Same comments as for the slow tokenizer (I think it will just need a copy-paste once you apply the suggestions in the slow tokenizer).

**Review by sgugger — APPROVED** [review]
This is very clean! Just have a few last nits on the docstrings of the tokenizers (and as Lysandre and Patrick mentioned, we need the test modeling files to merge this).
Thanks a lot for all your work!

**Comment by sgugger** [comment]
Oh and another thing to do after the merge will be to add your new model to the main README and the documentation so that people can use it! The template should give you a file for the `.rst` (or you can use `docs/model_doc/bert.rst` as an example).

**Review by jplu — APPROVED** [review]
Thanks a lot!! This is a very good work! I'm totally ok with the TF part once there will be a test file added.

File: src/transformers/models/mpnet/tokenization_mpnet.py
Comment (StillKeepTry) [reply src/transformers/models/mpnet/tokenization_mpnet.py]: @patrickvonplaten I have removed inheritance and updated [tokenization_mpnet.py](https://github.com/StillKeepTry/transformers/blob/master/src/transformers/models/mpnet/tokenization_mpnet.py) and [tokenization_mpnet_fast.py](https://github.com/StillKeepTry/transformers/blob/master/src/transformers/models/mpnet/tokenization_mpnet_fast.py).  Can you have a check again?

**Comment by StillKeepTry** [comment]
I have updated `test_modeling_mpnet.py` now.

**Comment by StillKeepTry** [comment]
Hi, every reviewer. I have fixed previous comments (like doc, format, and so on) and updated the `tokenization_mpnet.py` and `tokenization_mpnet_fast.py` by removing the inheritance. Besides, I also update `test_modeling_mpnet.py`, `test_modeling_tf_mpnet.py` for testing, and model weights into the model hub. 

So can you make some new reviews based on the current version? Thank you very much.

**Comment by StillKeepTry** [comment]
Hi, every reviewer. Thank you for your valuable reviews. I have fixed previous comments (like doc, format, and so on) and updated the `tokenization_mpnet.py` and `tokenization_mpnet_fast.py` by removing the inheritance. Besides, I also upload test files (`test_modeling_mpnet.py`, `test_modeling_tf_mpnet.py`) for testing, and model weights into the model hub.

**Comment by LysandreJik** [comment]
Fantastic, thanks for working on it! Will review today.

**Comment by StillKeepTry** [comment]
@patrickvonplaten Hi, are there any new comments?

**Comment by jplu** [comment]
Hello!

Still some comments:

1. Update the inputs handling in the TF file, we have merged an update for the booleans. You can see an example in the TF BERT file if you need one.
2. rebase and fix the conflicting files.
3. Fix the check_code_quality test.

**Review by patrickvonplaten — APPROVED** [review]
LGTM, if possible (if there are pre-trained weights) it would be great to also add an integration test with a pre-trained weight, similar to how it's done for roberta here: https://github.com/huggingface/transformers/blob/28c77ddf3bd39f3e528f41d94a67b88ad967173a/tests/test_modeling_roberta.py#L426.

**Comment by patrickvonplaten** [comment]
I think something went wrong with the merge here :-/ Could you try to open a new PR that does not include all previous commits or fix this one?

**Comment by StillKeepTry** [comment]
> I think something went wrong with the merge here :-/ Could you try to open a new PR that does not include all previous commits or fix this one?

OK, it seems something wrong when I update to the latest version. :(

**Comment by StillKeepTry** [comment]
@patrickvonplaten @JetRunner @sgugger @LysandreJik @jplu The new PR is moved to [https://github.com/huggingface/transformers/pull/8971](https://github.com/huggingface/transformers/pull/8971)

**Pull request closed without merging**
